Feature: Change reload to be configurable with glob patterns

Roang-zero1 opened the pull request:
With this change it is possible to control which files will trigger a reload of uvicorn.

Features:

- New Option `reload-include` is a repeatable list of unix style glob patterns [pathlib.Path.glob](https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob) that will be included in reload watch.
- New Option `reload-exclude` is a repeatable list of unix style glob patterns [pathlib.Path.glob](https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob) that will be excluded in reload watch.
- Common Watchgod and Startreload configuration:
  - Both reloaders get their directories from the new configurations
- Improved Watchgod implementation:
  - Use `should_watch_dir` to reduce setup overhead
  - Use `should_watch_file` with `pathlib` to verify included paths
  - Only watch .py files by default
- Improved Startreload implementation:
  - Use pathlib glob to search for .py files
- Updated documentation
- Updated tests


Edit by @Kludex:
- Closes #984 
- Closes #966

Roang-zero1 committed:
Add tests for include and exclude

Roang-zero1 committed:
Add configuration for include and exclude

Roang-zero1 committed:
Update startreload

* Use Path instead of os.path for all manipulations
* Use glob pattern to search for python files

Roang-zero1 committed:
Update watchgodreload

* Use should_watch_dir to filter directories
* Use glob patterns to filter files

Roang-zero1 committed:
Update documentation

Roang-zero1 commented on uvicorn/supervisors/watchgodreload.py:
```diff
@@ -1,54 +1,72 @@
-    ignore_dotted_file_regex = r"^\/?(?:\w+\/)*(\.\w+)"
-    ignored = []
+    def __init__(self, root_path: Path, config: Config):
+        default_includes = ["*.py"]
```
Currently only .py files are watched, this is different to the current default behavior.
Please provide feedback which behavior is desired.

euri10 replied:
I'm ok with that, that would rquire a word or two in the changelog though

Roang-zero1 commented on uvicorn/supervisors/basereload.py:
```diff
@@ -16,8 +17,8 @@
 class BaseReload:
-    def __init__(self, config, target, sockets):
-        self.config = config
+    def __init__(self, config: Config, target, sockets):
```
I've added the type to this class as it is easier to develop on the inheriting classes.
Can be removed if desired.

Roang-zero1 commented on uvicorn/supervisors/statreload.py:
```diff
@@ -11,31 +11,31 @@ class StatReload(BaseReload):
+            old_time = self.mtimes.get(file)
             if old_time is None:
-                self.mtimes[filename] = mtime
+                self.mtimes[file] = mtime
```
Here the dict is now bound to a Path instead of a string.
I don't have any preference either way, so please indicate which way to go.

Roang-zero1 replied:
I've updated the mtimes to be a path in the basereload, I don't think a string is appropriate anymore.

euri10 commented on docs/settings.md:
```diff
@@ -19,9 +19,11 @@ equivalent keyword arguments, eg. `uvicorn.run("example:app", port=5000, reload=
+* `--reload-exclude <glob-pattern>` - Specify a glob pattern to match files or directories which will excluded from watching. May be used multiple times. By default the following patterns are excluded: `.*, .py[cod], .sw.*, ~*`. These can be overwritten by explicitly including them.
 
-By default Uvicorn uses simple changes detection strategy that compares python files modification times few times a second. If this approach doesn't work for your project (eg. because of its complexity), you can install Uvicorn with optional `watchgod` dependency to use filesystem events instead:
+By default Uvicorn uses simple changes detection strategy that compares python files modification times few times a second. If this approach doesn't work for your project (eg. because of its complexity), or you need watching of non python files, you can install Uvicorn with optional `watchgod` dependency to use filesystem events instead:
```
ok reading this I realize this is completely obsolete since we changed the packaging, so it needs to be removed.

```suggestion
By default Uvicorn uses simple changes detection strategy that compares python files modification times few times a second. If this approach doesn't work for your project (eg. because of its complexity), or you need watching of non python files
```

euri10 commented on docs/settings.md:
```diff
@@ -19,9 +19,11 @@ equivalent keyword arguments, eg. `uvicorn.run("example:app", port=5000, reload=
-* `--reload-dir <path>` - Specify which directories to watch for python file changes. May be used multiple times. If unused, then by default all directories in current directory will be watched.
+* `--reload-dir <path>` - Specify which directories to watch for python file changes. May be used multiple times. If unused, then by default the whole current directory will be watched.
+* `--reload-include <glob-pattern>` - Specify a glob pattern to match files or directories which will be watched. May be used multiple times. By default the following patterns are included: `.py`. These can be overwritten by explicitly excluding them.
+* `--reload-exclude <glob-pattern>` - Specify a glob pattern to match files or directories which will excluded from watching. May be used multiple times. By default the following patterns are excluded: `.*, .py[cod], .sw.*, ~*`. These can be overwritten by explicitly including them.
```
these are watchgod defaults as far as I know, in any case I'm not sure I'd enter that level of details in the docs unless there's a reason for it.

I recall we discussed at length this in another issue,

what happens in  the case no pattern is specified, which will be the most common use-case ?

Roang-zero1 replied:
This is in the docs in order to show which files are ignored. These ignores can be explicitly overwritten with an include.
I could also use super(), but then there would be no way to override the defaults.

Roang-zero1 replied:
@euri10  Should we remove this from the docs or do we want to keep it in there to let users know about these patterns?

euri10 replied:
I like it to have it here as well as in the cli `--help` output

Roang-zero1 replied:
Help text for include:
> Set glob patterns to include while watching
> for files. Includes '*.py' by default, which
> can be overridden in reload-excludes.

Help text for exclude:
>Set glob patterns to exclude while watching
>for files. Includes '.*, .py[cod], .sw.*,
>~*' by default, which can be overridden in
>reload-excludes.

Note I do not want to use the default parameter as it gives the impression that changing this will not include these patterns. Contrary to many other settings these patterns need to be explicitly negated.

euri10 commented on tests/supervisors/test_statreload.py:
```diff
@@ -11,37 +9,87 @@ def run(sockets):
+        reloader.run()
+
+
+class TestWatchGodReload(TestBaseReload):
```
```suggestion
class TestStatReload(TestBaseReload):
```

euri10 replied:
```suggestion
class TestStatReload(TestBaseReload):
```

euri10 commented on tests/supervisors/test_statreload.py:
```diff
@@ -11,37 +9,87 @@ def run(sockets):
 
 
-def test_statreload():
+def test_statreload(tmpdir):
```
why this test is left out of the refactor ?

euri10 commented on docs/settings.md:
```diff
@@ -19,9 +19,11 @@ equivalent keyword arguments, eg. `uvicorn.run("example:app", port=5000, reload=
 * `--reload` - Enable auto-reload.
-* `--reload-dir <path>` - Specify which directories to watch for python file changes. May be used multiple times. If unused, then by default all directories in current directory will be watched.
+* `--reload-dir <path>` - Specify which directories to watch for python file changes. May be used multiple times. If unused, then by default the whole current directory will be watched.
+* `--reload-include <glob-pattern>` - Specify a glob pattern to match files or directories which will be watched. May be used multiple times. By default the following patterns are included: `.py`. These can be overwritten by explicitly excluding them.
```
shouldn't it be *.py ?

```suggestion
* `--reload-include <glob-pattern>` - Specify a glob pattern to match files or directories which will be watched. May be used multiple times. By default the following patterns are included: `*.py`. These can be overwritten by explicitly excluding them.
```

euri10 commented on docs/settings.md:
````diff
@@ -19,9 +19,11 @@ equivalent keyword arguments, eg. `uvicorn.run("example:app", port=5000, reload=
+By default Uvicorn uses simple changes detection strategy that compares python files modification times few times a second. If this approach doesn't work for your project (eg. because of its complexity), or you need watching of non python files, you can install Uvicorn with optional `watchgod` dependency to use filesystem events instead:
 
 ```
 $ pip install uvicorn[watchgodreload]
````
```suggestion

```

euri10 replied:
this one still needs to be removed from the docs, it's not relevant anymore

edit: I mean the line `$ pip install uvicorn[watchgodreload]`

euri10 commented on docs/deployment.md:
```diff
@@ -36,6 +36,10 @@ Options:
   --reload                        Enable auto-reload.
   --reload-dir TEXT               Set reload directories explicitly, instead
                                   of using the current working directory.
+  --reload-include TEXT           Set glob patterns to exclude files in
```
```suggestion
  --reload-include TEXT           Set glob patterns to include files in
```

euri10 commented on docs/index.md:
```diff
@@ -105,6 +105,10 @@ Options:
   --reload                        Enable auto-reload.
   --reload-dir TEXT               Set reload directories explicitly, instead
                                   of using the current working directory.
+  --reload-include TEXT           Set glob patterns to exclude files in
```
```suggestion
  --reload-include TEXT           Set glob patterns to include files in
```

euri10 commented on tests/supervisors/test_watchgodreload.py:
```diff
@@ -12,96 +10,196 @@ def run(sockets):
 
 def test_watchgodreload(
-    tls_ca_certificate_pem_path, tls_ca_certificate_private_key_path
+    tls_ca_certificate_pem_path, tls_ca_certificate_private_key_path, tmpdir
```
could throw another PR to keep things separated but since it's not used here we should drop it
```suggestion
 tmpdir
```

euri10 commented on tests/supervisors/test_watchgodreload.py:
```diff
@@ -12,96 +10,196 @@ def run(sockets):
+
+            reloader.shutdown()
+
+    def test__not_should_reload_when_python_file_in_subdir_is_changed(self):
```
```suggestion
    def test_should_not_reload_when_python_file_in_subdir_is_changed(self):
```

euri10 reviewed (commented):
thanks for this !

I begun reviewing this and realize it would be simpler to split it in 2:

1. the test refactor
2. the added functionality

in no particular order, this one seems simpler from my point of view.

Sorry but cant wrap my head around reviewing that much to be honest, and in any case it's a separate issue.

I left my comments anyway since I begun but the split would be much appreciated

Roang-zero1 commented:
Sure, I'll start the work on separating out the test refactor.
I'll keep this PR for the reimplementation of the reload function and start a new one for the test refactor.

Roang-zero1 committed:
Apply suggestions from code review

Fix copy paste errors

Co-authored-by: euri10 <euri10@users.noreply.github.com>

Roang-zero1 committed:
Update docs/settings.md

Fix docs error

Co-authored-by: euri10 <euri10@users.noreply.github.com>

Roang-zero1 committed:
Update reload test class

* Use per test parameter to enable in depth testing

Roang-zero1 committed:
Update docs/settings.md

Co-authored-by: euri10 <euri10@users.noreply.github.com>

Roang-zero1 reviewed (commented):
Here are some points of consideration

Roang-zero1 commented:
As requested I've separated out the test class refactor to https://github.com/encode/uvicorn/pull/833.

Ready for new round of review. Added some comments to point out where feedback is requesed.

euri10 commented:
Will look at it over the weekend and get back to you, thanks !

On Fri, Oct 23, 2020, 7:22 PM Lucas <notifications@github.com> wrote:

> As requested I've separated out the test class refactor to #833
> <https://github.com/encode/uvicorn/pull/833>.
>
> Ready for new round of review. Added some comments to point out where
> feedback is requesed.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/encode/uvicorn/pull/820#issuecomment-715471940>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINSPX5VCEQNWOXUTB4SMDSMG3W7ANCNFSM4ST3ARHA>
> .
>

hanneskuettner commented:
Any updates on this PR?

Roang-zero1 commented:
Just got the notification. I would be willing to revisit this PR, just let me know what is needed.

Kludex commented:
Hi @Roang-zero1 :wave: 

I'm going to review it over the weekend. If you'd like, merging master would be great! :smile: 

Sorry for the delay on it!

kennylajara commented:
> I'm going to review it over the weekend. 

Bump @Kludex

Roang-zero1 committed:
Add test for StartReload output

Roang-zero1 committed:
Update types

Roang-zero1 committed:
Add coverage notation for TYPE_CHECKING

Roang-zero1 committed:
Fix documentation linting

Roang-zero1 commented:
I've rebased the changes onto the current master.
Unfortunately I don't know why the pipeline fails. I've run the lint and committed the changes, but no change.
Test Suite is successful on my local machine.

euri10 commented:
hello @Roang-zero1 , you just need to run `scripts/lint` : that will automatically update the docs with the latest changes you made in the click options

Roang-zero1 commented:
@euri10 That's what I did in fd185c4c2501ad3b2156172c28633d887f1cfc7c when I run it now there are no changes to commit.

The error is not present when I run scripts/check locally

euri10 commented:
I'd reinstall the venv
Maybe you run an old click version ?

euri10 commented:
I think that could be the reason the lint script behaves differently locally and in CI, iirc click 8 changed the cli help output format, so we need to be sure you have locally same version as in CI

if that's not the case I'm at a loss

Roang-zero1 committed:
Fix documentation for linting

Roang-zero1 commented:
Typical issue of using the wrong venv... It's fixed now thanks!

euri10 committed:
Removed tmpdir fixture

euri10 commented on tests/supervisors/test_reload.py:
```diff
@@ -10,12 +10,11 @@
     tmp_path: Path
 
     @pytest.fixture(autouse=True)
-    def setup(self, tmpdir, reloader_class):
```
I'm gonna be annoying, but I'd prefer we drop `tmpdir` entirely, see the small change I made on my branch (sent you a PR on your feat branch if this helps), we therefore deal only with the `tmp_path` fixture which is a nice `Path` object and just have to write a small context manager to change the directory in tests, seems cleaner and more explicit imho.

Roang-zero1 replied:
Nice, I could not figure out how to do that cleanly, so I reverted to tmpdir.
Will merge the branch.

euri10 committed:
Changed context manager to take path arg

euri10 commented on tests/test_config.py:
```diff
@@ -80,9 +80,36 @@ def test_config_should_reload_is_set(
+def test_non_existant_reload_dir_is_not_set(
+    tmpdir, caplog: pytest.LogCaptureFixture
+) -> None:
+    with tmpdir.as_cwd():
```
in light of the previous comment, maybe move the context manager as a test util and use it here too

euri10 commented on uvicorn/config.py:
```diff
@@ -224,13 +227,57 @@ def __init__(
+            try:
+                if Path(pattern).is_dir():
+                    reload_dirs_list.append(pattern)
+                    reload_includes.remove(pattern)
```
do we cover this ? I failed to see where 

ok sorry it's done

Roang-zero1 replied:
I've added a specific test case for just these config lines.
Thanks for making me write this test case I discovered a rather nasty bug during testing it (Pro Tip: never modify a list you are currently iterating 😅 )
See: 4bd5fd10d700c0a34856e532bd1c8e638a34ae96

euri10 commented on uvicorn/config.py:
```diff
@@ -224,13 +227,57 @@ def __init__(
+                if reload_paths[j] in reload_paths[k].parents:
+                    children.append(reload_paths[k])
+                elif reload_paths[k] in reload_paths[j].parents:
+                    children.append(reload_paths[j])
```
seems like 278 is not covered,
if I do a 
```
        children = []
        for j in range(len(reload_paths)):
            print(0)
            for k in range(j + 1, len(reload_paths)):
                if reload_paths[j] in reload_paths[k].parents:
                    print(1)
                    children.append(reload_paths[k])
                elif reload_paths[k] in reload_paths[j].parents:
                    print(2)
                    children.append(reload_paths[j])
```
and run the test suite it seems no 1 or 2 are printed, I'm not sure I get what this children is doing, would you show a test case that demonstrate its use ?

Roang-zero1 replied:
The use case is to find directories which are children of other watched directories.
- app
  - sub

and both app and sub are added to reloaddirs.
It would be a waste to create watchers for both app and sub, therefore I remove the children.
I have added a [new test case](https://github.com/Roang-zero1/uvicorn/commit/9a10bf80f7077ea9e78ca7cefe0d0217746a9ae6#diff-da24b355349b53e4c8047034767d008e4c2ce9ea6c8933a3acd50d0828f00528R132-R151), unfortunately I am [using a set](https://github.com/Roang-zero1/uvicorn/blob/9a10bf80f7077ea9e78ca7cefe0d0217746a9ae6/uvicorn/config.py#L255-L260) beforehand so I can't guarantee that both branches are met every time...
@euri10 Do you have input on how I could solve this in the test?

euri10 replied:
no this looks fine this way

euri10 commented on uvicorn/supervisors/watchgodreload.py:
```diff
@@ -12,23 +10,59 @@
+                for exclude_pattern in self.excludes:
+                    if entry_path.match(exclude_pattern):
+                        return False
+                return True
```
that's beautiful

euri10 commented:
I made a few comments @Roang-zero1 but this looks good !

Roang-zero1 committed:
Merge pull request #2 from euri10/Roang-zero1-feature/reload-exclude-include

Removed tmpdir fixture

Roang-zero1 committed:
Update config tests with as_cwd util

* Move as_cwd to test utils
* Update test_config tests with new util
* Remove usage of tmpdir

Roang-zero1 committed:
Remove debug print statement

Roang-zero1 committed:
Fix reload_includes arbitrarliy dropping entries

Roang-zero1 committed:
Add additional configuration tests

Roang-zero1 committed:
Fix mypy error for mkdir

euri10 commented on tests/supervisors/test_reload.py:
```diff
@@ -96,3 +170,88 @@ def test_should_reload_when_directories_have_same_prefix(self):
+            reloader.shutdown()
+
+    @pytest.mark.parametrize("reloader_class", [StatReload])
+    def test_should_not_parse_filetype_from_includes(self) -> None:
```
this  one test looks weird to me or I dont get the intent : you include all js files to be checked for a change, then assert that it's not reloaded, shouldn't it be the opposite ?

Roang-zero1 replied:
See my latest comment. This is more to the functionality as a whole that should be discussed.

euri10 reviewed (commented):
small comment and we soon gtg !

Roang-zero1 commented:
There is a conceptional question I just realized must be discussed.
There are currently 3 settings:
- reload_dir a list setting where one can explicitly set directories which should be watched.
- reload_include a list of glob patterns to match files which trigger a reload.
- reload_exclude a list of glob patterns to match for files which to not trigger a reload.

As the pull request currently stands any directory in reload_dir and any pattern that exactly matches a directory in reload_includes is matched and added to reload_dir. This means it is currently not possible to glob for directories.
I think I should change this to be possible as it would be against my intuition when using these settings.

Example:
Directory structure
root
├── app_first
├── app_second
├── external
├── javascript
└── stylesheets

Let's say our user chose to include `javascript` and `stylesheets` in reload_dir  but decided to use a glob for the apps `app_*`. Then currently the app directories would *not* be watched. I think this is an problem.

I have four ways to resolve this:
- Separate the functionality of reload_dir and reload_include. Therefore the user has to explicitly set the directories they want to match.
- Resolve the directory glob patterns on startup and add them to the reload_dis if they are directories (Would increase startup costs, but is one time)
- Do not resolve the directory globs and let the `should_watch_dir` of `watchgod` handle the check if it is included
- A hybrid of 2 and 3 where the globs are resolved at the start, but watchgod still verifies the patterns against the triggered directories.

I would vote for the fourth solution. Although it adds complexity it is a good middle ground between speed and functionality.

Roang-zero1 committed:
Remove installation hint for watchgod

Co-authored-by: euri10 <euri10@users.noreply.github.com>

euri10 commented:
I'm also in favor of option `#4` @Roang-zero1 and I agree globs should include directories

Roang-zero1 committed:
➕ Add warning for invalid reload config

If any of the reload settings are present, but the app would not reload
there will not be a warning informing the user.

Roang-zero1 committed:
➕ Add reload_dir result as log message

Roang-zero1 committed:
➕ Add reload_directory_structure fixture

Use one directory structure for all reload test.

Roang-zero1 committed:
➕ Add reload defaults to cli help

Roang-zero1 committed:
➕ Add parsing of reload dirs from glob patterns

Glob patterns are now matched agains directories which will
be added to reload_dirs if not present.
This also extends to excluded directories.
Exclude patterns now take precedence over includes.

Roang-zero1 committed:
🐛 Fix incorrect init in StatReload

Roang-zero1 committed:
🔨 Refactor should_watch_dir for WatchGodReload

Uses a global dict to cache results.
Prioritizes exclude directories and patterns.
Finds additional reload/exclude directories if they are created under an
existing watcher.

Roang-zero1 committed:
➕ Add result cache for should_watch_*

Roang-zero1 committed:
🐛 Fix slow first reload test

First reload test was working without the prepared directory.
Therefore it initialized on the repository itself, thereby being slow.

Roang-zero1 commented:
I've added the changes, unfortunately this meant reworking large parts of this PR 🔨. (Sorry @euri10 for the additional work)
There is now only one watcher for the current working directory. This enables watchgod to pick up new files and directories that match the globs that were not present before starting the server.

I think I've added test cases for all new functionality.
I am a bit on the fence as there is some feature duplication between `config.py` and the `watchgodreload.py`, but I believe this to be unavoidable.

I've added a cache to the watchers, this reduces the CPU load, but costs memory. Please indicate if that is ok.

There is now a fixture for testing all the reload related use cases, it is currently scoped for `package`, as I don't have a good enough understanding please verify it the scope is correct.

Roang-zero1 commented:
Just saw the failed test. I'll look into what the problem with Windows is tomorrow.

Roang-zero1 committed:
🐛 Fix directory checks for Windows

Windows raises an OSError when a path is not found.
This change introduces a helper function to mitigate this.

Roang-zero1 committed:
🐛 Fix reload warning message tests

Windows paths were incorrect in the tests, this fixes these messages.

euri10 commented:
this is a subtle one:
this test works fine with this logger: `logger.info(f"Will watch for changes in these directories: {list(map(str, self.reload_dirs))}")`
```
        assert (
            caplog.records[-1].message
            == f"Will watch for changes in these directories: {[str(app_dir)]}"
        )
```
you cannot build manually the list `[]`  I think that was what caused the double escaping

euri10 commented on tests/conftest.py:
```diff
@@ -71,3 +71,43 @@ def tls_ca_ssl_context(tls_certificate_authority: trustme.CA) -> ssl.SSLContext:
+
+@pytest.fixture(scope="package")
+def reload_directory_structure(tmp_path_factory: pytest.TempPathFactory):
+    """This fixture creates a directory structure to enable reload parameter tests"""
```
```suggestion
    """
    This fixture creates a directory structure to enable reload parameter tests
    
    ├── app
    │ ├── css
    │ │ └── main.css
    │ ├── js
    │ │ └── main.js
    │ ├── src
    │ │ └── main.py
    │ └── sub
    │     └── sub.py
    ├── app_first
    │ ├── css
    │ │ └── main.css
    │ ├── js
    │ │ └── main.js
    │ ├── src
    │ │ └── main.py
    │ └── sub
    │     └── sub.py
    ├── app_second
    │ ├── css
    │ │ └── main.css
    │ ├── js
    │ │ └── main.js
    │ ├── src
    │ │ └── main.py
    │ └── sub
    │     └── sub.py
    ├── app_third
    │ ├── css
    │ │ └── main.css
    │ ├── js
    │ │ └── main.js
    │ ├── src
    │ │ └── main.py
    │ └── sub
    │     └── sub.py
    ├── ext
    │ └── ext.jpg
    └── main.py
    """
```

euri10 replied:
not sure about that suggestion, just something I may want to see because reading the test it was not obvious from me reading the fixture what the directory layout looked like, but this is very long so....

Roang-zero1 replied:
I've taken the middle road and just shortened the directories down to one example with indication they are all the same

euri10 commented on tests/supervisors/test_reload.py:
```diff
@@ -36,63 +39,354 @@ def _reload_tester(self, reloader: BaseReload, file: Path) -> bool:
+            reloader.shutdown()
+
+    @pytest.mark.parametrize("reloader_class", [WatchGodReload])
+    def test_should_not_reload_when_python_file_in_subdir_is_changed(self) -> None:
```
```suggestion
    def test_should_not_reload_when_python_file_in_excluded_subdir_is_changed(self) -> None:
```

euri10 commented on tests/supervisors/test_reload.py:
```diff
@@ -36,63 +39,354 @@ def _reload_tester(self, reloader: BaseReload, file: Path) -> bool:
+
+            reloader.shutdown()
+
+    @pytest.mark.parametrize("reloader_class", [StatReload])
```
ok here I dont follow, I guess  that will be different here for the `WatchGodReload` case ? why not doing the same as in `test_reload_when_pattern_matched_file_is_changed` or put differently, what does this test differently than `test_reload_when_pattern_matched_file_is_changed`

Roang-zero1 replied:
Sorry Github is not showing the test name, which one is this comment for?

euri10 replied:
for     `def test_should_not_parse_filetype_from_includes(self) -> None:`

Roang-zero1 replied:
Ok, it's an old duplicate test. I have removed it

euri10 commented on tests/supervisors/test_reload.py:
```diff
@@ -36,63 +39,354 @@ def _reload_tester(self, reloader: BaseReload, file: Path) -> bool:
-
             assert self._reload_tester(reloader, app_file)
-            assert self._reload_tester(reloader, app_ext_file)
+
+            app_first_dir.mkdir()
+            assert self._reload_tester(reloader, app_first_file)
+            assert caplog.records[-2].levelno == INFO
+            assert (
+                caplog.records[-2].message == "WatchGodReload detected a new reload "
+                f"dir '{app_first_dir.name}' in '{tmp_path}'; Adding to watch list."
+            )
+
+            reloader.shutdown()
```
this is neat

euri10 commented on tests/test_config.py:
```diff
@@ -80,9 +81,156 @@ def test_config_should_reload_is_set(
+        assert len(caplog.records) == 1
+        assert (
+            caplog.records[-1].message
+            == f"Will watch for changes in these directories: ['{app_dir}']"
```
see my comment about not using `[]`

euri10 commented on tests/test_config.py:
```diff
@@ -80,9 +81,156 @@ def test_config_should_reload_is_set(
+        )
+
+
+def test_reload_subdir_removal(reload_directory_structure: Path) -> None:
```
```suggestion
def test_duplicate_direcories_are_removed(reload_directory_structure: Path) -> None:
```
not sure about this one I let you choose

euri10 commented on uvicorn/config.py:
```diff
@@ -132,6 +133,57 @@ def create_ssl_context(
+    patterns_list: List[str], directories_list: List[str]
+) -> Tuple[List[str], List[Path]]:
+
+    directories: List[Path] = list(set(map(Path, directories_list.copy())))
```
writing this without thinking too much or having read the rest, would a deepcopy be better or not ?

Roang-zero1 replied:
No we don't need a deep copy here as we can keep the path elements. I just wanted to decouple the list inside from the outside as I still need the outside one to check what was passed to the config.

euri10 commented on uvicorn/config.py:
```diff
@@ -132,6 +133,57 @@ def create_ssl_context(
+        if is_dir(Path(pattern)):
+            directories.append(Path(pattern))
+        else:
+            print("here")
```
let's remove that

Roang-zero1 replied:
Removed

euri10 commented on uvicorn/supervisors/basereload.py:
```diff
@@ -64,7 +64,6 @@ def startup(self) -> None:
         self.process.start()
 
     def restart(self) -> None:
-        self.mtimes: Dict[str, float] = {}
```
why this is dropped ?

Roang-zero1 replied:
It was moved to StartReload as it is not used anywhere else

euri10 commented on uvicorn/supervisors/statreload.py:
```diff
@@ -19,31 +18,35 @@ def __init__(
         super().__init__(config, target, sockets)
         self.reloader_name = "statreload"
-        self.mtimes = {}
+        self.mtimes: Dict[Path, float] = {}
```
because you put it here ?

Roang-zero1 replied:
Correct

euri10 commented on uvicorn/supervisors/watchgodreload.py:
```diff
@@ -12,23 +10,107 @@
 
     def should_watch_file(self, entry: "DirEntry") -> bool:
-        return not any(r.search(entry.name) for r in self._ignored)
+        cached_result = self.watched_files.get(entry.path)
+        if cached_result is not None:
+            return cached_result
```
is it made to be faster ?

Roang-zero1 replied:
Yes, depending on the amount of patterns it might take a while to match all the files. I had to implement it for the directories anyway and reused it for the files.
For the files the cache is something I am on the fence about...

Roang-zero1 committed:
Update tests/supervisors/test_reload.py

Co-authored-by: euri10 <euri10@users.noreply.github.com>

Roang-zero1 committed:
🔥 Remove debug print statement

Roang-zero1 committed:
🚨 Fix test linting

Roang-zero1 committed:
➕ Add description for reload_directory_structure fixture

Roang-zero1 committed:
🐛 Fix path error in windows tests

Roang-zero1 committed:
🐛 Fix path test again

euri10 commented:
for the failing test, let's deal with that sorting the files in the output for predictable behaviour:

```diff
diff --git a/tests/test_config.py b/tests/test_config.py
index a999b33..73f2e5f 100644
--- a/tests/test_config.py
+++ b/tests/test_config.py
@@ -225,8 +225,10 @@ def test_reload_includes_exclude_dir_patterns_are_matched(
                 "Will watch for changes in these directories: "
                 in caplog.records[-1].message
             )
-            assert str(first_app_dir) in caplog.records[-1].message
-            assert str(second_app_dir) in caplog.records[-1].message
+            assert (
+                    caplog.records[-1].message
+                    == f"Will watch for changes in these directories: {sorted([str(first_app_dir), str(second_app_dir)])}"
+            )
             assert frozenset(config.reload_dirs) == frozenset(
                 [first_app_dir, second_app_dir]
             )

             )
             )
diff --git a/uvicorn/config.py b/uvicorn/config.py
index 4fae430..13f5898 100644
--- a/uvicorn/config.py
+++ b/uvicorn/config.py
@@ -331,7 +331,7 @@ class Config:

             logger.info(
                 "Will watch for changes in these directories: %s",
-                list(map(str, self.reload_dirs)),
+                sorted(list(map(str, self.reload_dirs))),
             )

         if env_file is not None:
```

Roang-zero1 committed:
🔥 Remove duplicate test case

Roang-zero1 committed:
➕ Add sorting to reload_dirs list

euri10 commented on uvicorn/supervisors/watchgodreload.py:
```diff
@@ -41,26 +123,14 @@ def __init__(
+        for directory in config.reload_dirs:
+            if Path.cwd() not in directory.parents:
+                reload_dirs.append(directory)
+        if Path.cwd() not in reload_dirs:
+            reload_dirs.append(Path.cwd())
```
I'm just wondering about those 2 lines because I was trying the branch on my ucivorn repo:
it has a dev folder I gitignored to contain all the dev app I'm loading to debug, and several dot files that are the result of mypy / pytest etc...
```
uvicorn
|--dev
|--.git
|--.idea
|--.pytest_cache
```
now if I do `dev.app:app --reload --log-level=trace --reload-dir=dev`
I get 
```
INFO:     Will watch for changes in these directories: ['/home/lotso/PycharmProjects/uvicorn/dev']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
DEBUG:    WatchGodReload detected a new excluded dir '.github' in '/home/lotso/PycharmProjects/uvicorn'; Adding to exclude list.
DEBUG:    WatchGodReload detected a new excluded dir '.git' in '/home/lotso/PycharmProjects/uvicorn'; Adding to exclude list.
DEBUG:    WatchGodReload detected a new excluded dir '.eggs' in '/home/lotso/PycharmProjects/uvicorn'; Adding to exclude list.
DEBUG:    WatchGodReload detected a new excluded dir '.idea' in '/home/lotso/PycharmProjects/uvicorn'; Adding to exclude list.
DEBUG:    WatchGodReload detected a new excluded dir '.pytest_cache' in '/home/lotso/PycharmProjects/uvicorn'; Adding to exclude list.
INFO:     Started reloader process [1445703] using watchgod
```

so looking for changes in dev is ok to me, the unexpected here is the dot folders at the same level.
not sure if that's intended

Roang-zero1 replied:
Yeah that's intended, but I see that it is not very intuitive. I'll add a check for the exclude warning.
What is happening is the following:

- We add the reload dir to folders we watch
- We need to add the cwd for folders that could be later found via an include patterns
- The cwd is scanned and foders that are excluded via the default pattern .* are found in cwd, that leads to this messages

I'll check if found folders are in the reload dirs before firing of this warning, that should clear up the unnecessary information.

Roang-zero1 replied:
I've added the check, but it will still log for the default patterns (esp. hidden directories) in watched directories.
I think this is fine as it is DEBUG level output...

Roang-zero1 committed:
🐛 Fix redundant excluded directory log output

Log messages were printed for excluded directories, even if these
directories were not in any of the watched reload_dirs.

euri10 commented:
thanks a lot @Roang-zero1 for this, this is a very nice enhancement !

HansBrende commented:
@Roang-zero1 @euri10 @Kludex this PR introduced a regression and deleted the test for it (`test_reload_dir_is_set()`). Previously you could pass `reload_dirs` a string instead of a list. Now:
```python
Config('myapp:app', reload=True, reload_dirs='src')
```
prints:
```
WARNING:  Provided reload directories ['r', 'c', 's'] did not contain valid directories, watching current working directory.
```
https://github.com/encode/uvicorn/pull/820/files#diff-6982a5e270879281f9182303a9da9776148857b46325afb230e50ef8eb3264c9R294
(Of course, `set` expects an iterable... and strings are iterables of chars.)
